src/Core/Services/DeviceCryptoService.cs

```diff
@@ -0,0 +1,85 @@
+﻿
+using System;
+using System.Threading.Tasks;
+using Bit.Core.Models.Domain;
+using Bit.Core.Models.Request;
+
+namespace Bit.Core.Abstractions
+{
+    public class DeviceCryptoService
+    {
+        private readonly IApiService _apiService;
+        private readonly IAppIdService _appIdService;
+        private readonly ICryptoFunctionService _cryptoFunctionService;
+        private readonly ICryptoService _cryptoService;
+        private readonly IStateService _stateService;
+
+        public DeviceCryptoService(
+            IApiService apiService,
+            IAppIdService appIdService,
+            ICryptoFunctionService cryptoFunctionService,
+            ICryptoService cryptoService,
+            IStateService stateService)
+        {
+            _apiService = apiService;
+            _appIdService = appIdService;
+            _cryptoFunctionService = cryptoFunctionService;
+            _cryptoService = cryptoService;
+            _stateService = stateService;
+        }
+
+
+        public async Task<DeviceResponse> TrustDeviceAsync()
+        {
+            // Attempt to get user key
+            var userKey = await _cryptoService.GetEncKeyAsync();
+            if (userKey == null)
+            {
+                return null;
+            }
+            // Generate deviceKey
+            var deviceKey = await this.MakeDeviceKeyAsync();
+
+            // Generate asymmetric RSA key pair: devicePrivateKey, devicePublicKey
+            var deviceKeyPair = await _cryptoFunctionService.RsaGenerateKeyPairAsync(2048);
+
+            var encryptUserKeyTask = _cryptoService.RsaEncryptAsync(userKey.EncKey, deviceKeyPair.Item1);
+            var encryptPublicKeyTask = _cryptoService.EncryptAsync(deviceKeyPair.Item1, userKey);
+            var encryptPrivateKeyTask = _cryptoService.EncryptAsync(deviceKeyPair.Item2, deviceKey);
+
+            // Send encrypted keys to server
+            var deviceIdentifier = await _appIdService.GetAppIdAsync();
+            var deviceRequest = new TrustedDeviceKeysRequest()
+            {
+                EncryptedUserKey = (await encryptUserKeyTask).EncryptedString,
+                EncryptedPublicKey = (await encryptPublicKeyTask).EncryptedString,
+                EncryptedPrivateKey = (await encryptPrivateKeyTask).EncryptedString,
+            };
+            return await _apiService.UpdateTrustedDeviceKeysAsync(deviceIdentifier, deviceRequest);
```

Comment:
⛏️ Remove the use of `this` (I thought this was caught by the formatting tool)

⛏️ IMO it's easier to read and understand what part is by deconstructing the tuple, so instead of `deviceKeyPair` you can have `var (devicePublicKey, devicePrivateKey)` and you can remove the comment given that it's more readable this way. And on the usage you can directly use `devicePublicKey` instead of `deviceKeyPair.Item1`.

🤔  Is there a reason to defer the encrypt task executions? Also you could run the three tasks in parallel using `Task.WhenAll(...)` and it gives back the results of all the operations. On the other hand, you can directly `await` on the spot, so instead of creating task variables you can have the awaited result there to be use afterwards in the device request.

⛏️ You can remove the parenthesis `()` from the `new TrustedDeviceKeyRequest` if you want as well.

Reply:
I was attempting to execute the tasks while we do the rest of the work, but its not worth it. Removed.